src/tactic/linear_combination.lean

```diff
@@ -0,0 +1,386 @@
+  if `hsum_on_left` is 5*x - 5*y = 0, and the target is -5*y + 5*x = 0, this
+  tactic will change the target to be -5*y + 5*x = 5*x - 5*y.
+
+This tactic only should be used when the target is an equality whose right side
+  is 0.
+
+* Input:
+  * `hsum_on_left` : expr, which should be an equality whose right side is 0
+
+* Output: tactic unit
+-/
+meta def set_goal_to_hleft_eq_tleft (hsum_on_left : expr) : tactic unit :=
+do
+  { change_goal ← tactic.to_expr ``(replace_eq_expr %%hsum_on_left),
+    tactic.apply change_goal,
+    pure () }
+  <|> tactic.fail ("The type of the left and right sides of each equality " ++
+    "must fulfill the 'has_zero' condition.")
+
+/--
+This tactic attempts to prove the goal by normalizing the target if the
+`normalize` field of the given configuration is true.
+
+* Input:
+  * `config` : a linear_combination_config, which determines the tactic used
+      for normalization if normalization is done
+
+* Outout: tactic unit
+-/
+meta def prove_equal_if_desired (config : linear_combination_config) :
+  tactic unit :=
+when config.normalize config.normalization_tactic
+
+/-! ### Part 4: Completed Tactic -/
+
+
+/--
+This is a tactic that attempts to prove the target by creating and applying a
+  linear combination of a list of equalities.  (If the `normalize` field of the
+  configuration is set to ff, then the tactic will simply set the user up to
+  prove their target using the linear combination instead of attempting to
+  finish the proof.)
+
+Note: The left and right sides of all the equations should have the same
+  ring type, and the coefficients should also have this type.  This type must
+  have the has_mul and add_group properties.  Also note that the target must
+  involve at least one variable.
+
+* Input:
+  * `heqs` : a list of names, referring to equations in the local context
+  * `coeffs` : a list of coefficients to be multiplied with the corresponding
+    equations in the list of names
+  * `config` : a linear_combination_config, which determines the tactic used
+      for normalization; by default, this value is the standard configuration
+      for a linear_combination_config
+
+* Output: tactic unit
+-/
+meta def linear_combination (heqs : list name) (coeffs : list pexpr)
+  (config : linear_combination_config := {}) : tactic unit :=
+do
+  hsum ← make_sum_of_hyps heqs coeffs,
+  hsum_on_left ← move_to_left_side hsum,
+  move_target_to_left_side,
+  set_goal_to_hleft_eq_tleft hsum_on_left,
+  prove_equal_if_desired config
+
+
+section interactive_mode
+setup_tactic_parser
+
+/--
+A parser that matches a pair in parentheses, where the first item in the pair
+is an identifier and the second item in the pair is a pexpr.
+
+* Input: None
+
+* Output: a lean.parser (name × pexpr)
+-/
+meta def parse_name_pexpr_pair : lean.parser (name × pexpr) :=
+do
+  tk "(",
+  id ← ident,
+  tk ",",
+  coeff ← parser.pexpr 0,
+  tk ")",
+  pure (id, coeff)
```

Comment:
These are my six lines of this PR :smiley: I find the expanded version more legible, personally, but it isn't important.

Reply:
I kind of like the fact that the one-line version has all the pieces arranged horizontally as in the resulting syntax. But really I wrote this only for didactic purposes since I guess @agoldb10 is maybe not so familiar with monadic combinators (see for instance [Mario's cheat sheet](https://leanprover-community.github.io/extras/tactic_writing.html#marios-monadic-symbols-cheat-sheet)).

Reply:
Thank you for showing me how else this could be written!  It is definitely helpful to know this syntax.  I think I will keep the expanded version if that is all right because I also find it to be more legible, but all of these versions are interesting.